File: internal/searchquery/query.go
// Package searchquery parses the small, portable search language exposed by
// symdesk. It deliberately does not generate SQL: callers can apply the plan
// to the sidecar while keeping parsing testable and shared across CLI, MCP and
// app clients.
package searchquery

import (
	"fmt"
	"regexp"
	"strings"
	"time"
	"unicode"
)

// Field is a metadata field supported by the query language.
type Field string

const (
	FieldPath     Field = "path"
	FieldTag      Field = "tag"
	FieldType     Field = "type"
	FieldStatus   Field = "status"
	FieldFilename Field = "filename"
	FieldFileType Field = "filetype"
	FieldCreated  Field = "created"
	FieldModified Field = "modified"
)

// Filter limits results using an indexed metadata field.
type Filter struct {
	Field   Field
	Value   string
	Negated bool
}

// Term is a full-text term. Phrases retain their exact multi-word boundary;
// unquoted terms are prefix matches, preserving the existing FTS behaviour.
type Term struct {
	Value   string
	Phrase  bool
	Negated bool
}

// Regex is applied to a complete document after the sidecar's indexed filters
// and FTS terms have narrowed the candidate set. Go's RE2 implementation keeps
// user expressions bounded and avoids catastrophic backtracking.
type Regex struct {
	Pattern string
	Negated bool
	re      *regexp.Regexp
}

// Matches reports whether the compiled regular expression matches text.
func (r Regex) Matches(text string) bool {
	return r.re.MatchString(text)
}

// Plan is the parsed query. All entries are combined with AND semantics.
type Plan struct {
	Filters []Filter
	Terms   []Term
	Regexes []Regex
}

// RequiresSidecar reports whether the query uses syntax that a sibling search
// tool cannot faithfully interpret. Plain terms may still use symseek.
func (p Plan) RequiresSidecar() bool {
	if len(p.Filters) > 0 || len(p.Regexes) > 0 {
		return true
	}
	for _, term := range p.Terms {
		if term.Phrase || term.Negated {
			return true
		}
	}
	return false
}

// Parse parses path:, tag:, type:, status:, filename:, filetype:, created:,
// modified:, quoted phrases, -negation and /regular expressions/. A syntax error
// is intentionally returned to callers;
// they can then degrade the whole original query to safe plain full-text search.
func Parse(input string) (Plan, error) {
	var plan Plan
	for i := 0; ; {
		i = skipSpace(input, i)
		if i >= len(input) {
			return plan, nil
		}

		negated := false
		if input[i] == '-' {
			negated = true
			i++
			if i >= len(input) || unicode.IsSpace(rune(input[i])) {
				return Plan{}, fmt.Errorf("negation must be followed by a term")
			}
		}

		switch input[i] {
		case '"':
			value, next, err := quoted(input, i)
			if err != nil {
				return Plan{}, err
			}
			plan.Terms = append(plan.Terms, Term{Value: value, Phrase: true, Negated: negated})
			i = next
		case '/':
			pattern, next, err := regex(input, i)
			if err != nil {
				return Plan{}, err
			}
			compiled, err := regexp.Compile(pattern)
			if err != nil {
				return Plan{}, fmt.Errorf("invalid regular expression: %w", err)
			}
			plan.Regexes = append(plan.Regexes, Regex{Pattern: pattern, Negated: negated, re: compiled})
			i = next
		default:
			value, next := bare(input, i)
			if value == "" {
				return Plan{}, fmt.Errorf("expected search term")
			}
			if field, filterValue, ok, err := filter(value); err != nil {
				return Plan{}, err
			} else if ok {
				// Relative date windows intentionally remain readable (`last week`)
				// while the rest of the grammar is whitespace-delimited. Consume
				// the unit as part of a date filter only.
				if isDateField(field) && strings.EqualFold(filterValue, "last") {
					unitStart := skipSpace(input, next)
					unit, unitEnd := bare(input, unitStart)
					if isDateUnit(unit) {
						filterValue += " " + strings.ToLower(unit)
						next = unitEnd
					}
				}
				if isDateField(field) {
					if err := ValidateDateValue(filterValue); err != nil {
						return Plan{}, err
					}
				}
				plan.Filters = append(plan.Filters, Filter{Field: field, Value: filterValue, Negated: negated})
			} else {
				plan.Terms = append(plan.Terms, Term{Value: value, Negated: negated})
			}
			i = next
		}

		if i < len(input) && !unicode.IsSpace(rune(input[i])) {
			return Plan{}, fmt.Errorf("unexpected character %q after search term", input[i])
		}
	}
}

func skipSpace(input string, i int) int {
	for i < len(input) && unicode.IsSpace(rune(input[i])) {
		i++
	}
	return i
}

func bare(input string, i int) (string, int) {
	start := i
	for i < len(input) && !unicode.IsSpace(rune(input[i])) {
		i++
	}
	return input[start:i], i
}

func quoted(input string, i int) (string, int, error) {
	start := i
	i++ // opening quote
	var value strings.Builder
	escaped := false
	for i < len(input) {
		ch := input[i]
		i++
		if escaped {
			value.WriteByte(ch)
			escaped = false
			continue
		}
		if ch == '\\' {
			escaped = true
			continue
		}
		if ch == '"' {
			if value.Len() == 0 {
				return "", 0, fmt.Errorf("empty quoted phrase at offset %d", start)
			}
			return value.String(), i, nil
		}
		value.WriteByte(ch)
	}
	return "", 0, fmt.Errorf("unterminated quoted phrase at offset %d", start)
}

func regex(input string, i int) (string, int, error) {
	start := i
	i++ // opening slash
	var pattern strings.Builder
	escaped := false
	for i < len(input) {
		ch := input[i]
		i++
		if escaped {
			pattern.WriteByte('\\')
			pattern.WriteByte(ch)
			escaped = false
			continue
		}
		if ch == '\\' {
			escaped = true
			continue
		}
		if ch == '/' {
			if pattern.Len() == 0 {
				return "", 0, fmt.Errorf("empty regular expression at offset %d", start)
			}
			return pattern.String(), i, nil
		}
		pattern.WriteByte(ch)
	}
	return "", 0, fmt.Errorf("unterminated regular expression at offset %d", start)
}

func filter(token string) (Field, string, bool, error) {
	fieldName, value, hasSeparator := strings.Cut(token, ":")
	if !hasSeparator {
		return "", "", false, nil
	}
	if value == "" {
		return "", "", false, fmt.Errorf("operator %q requires a value", fieldName+":")
	}

	switch strings.ToLower(fieldName) {
	case string(FieldPath):
		return FieldPath, value, true, nil
	case string(FieldTag):
		return FieldTag, value, true, nil
	case string(FieldType):
		return FieldType, value, true, nil
	case string(FieldStatus):
		return FieldStatus, value, true, nil
	case string(FieldFilename):
		return FieldFilename, value, true, nil
	case string(FieldFileType):
		return FieldFileType, value, true, nil
	case string(FieldCreated):
		return FieldCreated, value, true, nil
	case string(FieldModified):
		return FieldModified, value, true, nil
	default:
		if operatorName.MatchString(fieldName) {
			return "", "", false, fmt.Errorf("unknown search operator %q", fieldName+":")
		}
		return "", "", false, nil
	}
}

var operatorName = regexp.MustCompile(`^[[:alpha:]][[:alnum:]_-]*$`)

func isDateField(field Field) bool {
	return field == FieldCreated || field == FieldModified
}

func isDateUnit(value string) bool {
	switch strings.ToLower(value) {
	case "day", "week", "month", "year":
		return true
	default:
		return false
	}
}

// DateRange is an inclusive time interval used by created:/modified: filters.
type DateRange struct {
	From time.Time
	To   time.Time
}

// ParseDateValue resolves a date expression against reference. It accepts an
// ISO date, an RFC3339 timestamp, an inclusive `start..end` range, and the
// relative windows `last day`, `last week`, `last month`, and `last year`.
func ParseDateValue(value string, reference time.Time) (DateRange, error) {
	value = strings.TrimSpace(value)
	if value == "" {
		return DateRange{}, fmt.Errorf("date filter requires a value")
	}
	reference = reference.UTC()
	lower := strings.ToLower(value)
	if strings.HasPrefix(lower, "last ") {
		duration := map[string]time.Duration{
			"last day":   24 * time.Hour,
			"last week":  7 * 24 * time.Hour,
			"last month": 30 * 24 * time.Hour,
			"last year":  365 * 24 * time.Hour,
		}[lower]
		if duration == 0 {
			return DateRange{}, fmt.Errorf("invalid relative date %q", value)
		}
		return DateRange{From: reference.Add(-duration), To: reference}, nil
	}

	parts := strings.Split(value, "..")
	if len(parts) > 2 || (len(parts) == 2 && (strings.TrimSpace(parts[0]) == "" || strings.TrimSpace(parts[1]) == "")) {
		return DateRange{}, fmt.Errorf("invalid date range %q", value)
	}
	from, fromDate, err := parseDateEndpoint(strings.TrimSpace(parts[0]))
	if err != nil {
		return DateRange{}, err
	}
	to := from
	toDate := fromDate
	if len(parts) == 2 {
		to, toDate, err = parseDateEndpoint(strings.TrimSpace(parts[1]))
		if err != nil {
			return DateRange{}, err
		}
	}
	if fromDate {
		from = startOfUTCDate(from)
	}
	if toDate {
		to = startOfUTCDate(to).Add(24*time.Hour - time.Nanosecond)
	}
	if from.After(to) {
		return DateRange{}, fmt.Errorf("date range starts after it ends: %q", value)
	}
	return DateRange{From: from, To: to}, nil
}

// ValidateDateValue validates a date expression without making it depend on
// the wall clock. Relative expressions are checked against a fixed epoch.
func ValidateDateValue(value string) error {
	for _, expression := range strings.Split(value, ",") {
		if strings.TrimSpace(expression) == "" {
			return fmt.Errorf("invalid date %q", value)
		}
		if _, err := ParseDateValue(expression, time.Unix(0, 0).UTC()); err != nil {
			return err
		}
	}
	return nil
}

func parseDateEndpoint(value string) (time.Time, bool, error) {
	if t, err := time.Parse("2006-01-02", value); err == nil {
		return t.UTC(), true, nil
	}
	if t, err := time.Parse(time.RFC3339Nano, value); err == nil {
		return t.UTC(), false, nil
	}
	return time.Time{}, false, fmt.Errorf("invalid date %q (use YYYY-MM-DD or YYYY-MM-DD..YYYY-MM-DD)", value)
}

func startOfUTCDate(t time.Time) time.Time {
	u := t.UTC()
	return time.Date(u.Year(), u.Month(), u.Day(), 0, 0, 0, 0, time.UTC)
}

File: crates/symdesk-core/src/query.rs
#![deny(unsafe_code)]

//! Portable search-query grammar and inclusive UTC date filters.

use std::{error::Error, fmt};

use regex::Regex;
use time::{
    Date, Duration, OffsetDateTime, Time, format_description::well_known::Rfc3339,
    macros::format_description,
};

const DATE_FORMAT: &[time::format_description::FormatItem<'static>] =
    format_description!("[year]-[month]-[day]");

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Field {
    Path,
    Tag,
    Type,
    Status,
    Filename,
    FileType,
    Created,
    Modified,
    Index,
}

impl Field {
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Path => "path",
            Self::Tag => "tag",
            Self::Type => "type",
            Self::Status => "status",
            Self::Filename => "filename",
            Self::FileType => "filetype",
            Self::Created => "created",
            Self::Modified => "modified",
            Self::Index => "index",
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Filter {
    pub field: Field,
    pub value: String,
    pub negated: bool,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Term {
    pub value: String,
    pub phrase: bool,
    pub negated: bool,
}

#[derive(Clone, Debug)]
pub struct RegexTerm {
    pub pattern: String,
    pub negated: bool,
    compiled: Regex,
}

impl RegexTerm {
    #[must_use]
    pub fn matches(&self, text: &str) -> bool {
        self.compiled.is_match(text)
    }
}

#[derive(Clone, Debug, Default)]
pub struct Plan {
    pub filters: Vec<Filter>,
    pub terms: Vec<Term>,
    pub regexes: Vec<RegexTerm>,
}

impl Plan {
    #[must_use]
    pub fn requires_sidecar(&self) -> bool {
        !self.filters.is_empty()
            || !self.regexes.is_empty()
            || self.terms.iter().any(|term| term.phrase || term.negated)
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct QueryError {
    pub class: &'static str,
    pub message: String,
}

impl QueryError {
    fn new(class: &'static str, message: impl Into<String>) -> Self {
        Self {
            class,
            message: message.into(),
        }
    }
}

impl fmt::Display for QueryError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&self.message)
    }
}

impl Error for QueryError {}

/// Parses the Go-compatible query grammar.
///
/// # Errors
///
/// Returns a classified syntax/date/regex error without partially returning a plan.
pub fn parse(input: &str) -> Result<Plan, QueryError> {
    let bytes = input.as_bytes();
    let mut plan = Plan::default();
    let mut index = 0;
    loop {
        index = skip_space(bytes, index);
        if index >= bytes.len() {
            return Ok(plan);
        }
        let mut negated = false;
        if bytes[index] == b'-' {
            negated = true;
            index += 1;
            if index >= bytes.len() || is_go_space(bytes[index]) {
                return Err(QueryError::new(
                    "negation",
                    "negation must be followed by a term",
                ));
            }
        }

        match bytes[index] {
            b'"' => {
                let (value, next) = quoted(bytes, index)?;
                plan.terms.push(Term {
                    value,
                    phrase: true,
                    negated,
                });
                index = next;
            }
            b'/' => {
                let (pattern, next) = regex_literal(bytes, index)?;
                let rust_pattern = pattern.replace("\\/", "/");
                let compiled = Regex::new(&rust_pattern).map_err(|error| {
                    QueryError::new(
                        "invalid regular expression",
                        format!("invalid regular expression: {error}"),
                    )
                })?;
                plan.regexes.push(RegexTerm {
                    pattern,
                    negated,
                    compiled,
                });
                index = next;
            }
            _ => {
                let (value, next) = bare(bytes, index);
                if value.is_empty() {
                    return Err(QueryError::new(
                        "expected search term",
                        "expected search term",
                    ));
                }
                match parse_filter(&value)? {
                    Some((field, mut filter_value)) => {
                        let mut consumed_to = next;
                        if is_date_field(field) && filter_value.eq_ignore_ascii_case("last") {
                            let unit_start = skip_space(bytes, next);
                            let (unit, unit_end) = bare(bytes, unit_start);
                            if is_date_unit(&unit) {
                                filter_value.push(' ');
                                filter_value.push_str(&unit.to_lowercase());
                                consumed_to = unit_end;
                            }
                        }
                        if is_date_field(field) {
                            validate_date_value(&filter_value)?;
                        }
                        plan.filters.push(Filter {
                            field,
                            value: filter_value,
                            negated,
                        });
                        index = consumed_to;
                    }
                    None => {
                        plan.terms.push(Term {
                            value,
                            phrase: false,
                            negated,
                        });
                        index = next;
                    }
                }
            }
        }
        if index < bytes.len() && !is_go_space(bytes[index]) {
            return Err(QueryError::new(
                "unexpected character",
                format!(
                    "unexpected character {:?} after search term",
                    char::from(bytes[index])
                ),
            ));
        }
    }
}

fn quoted(input: &[u8], start: usize) -> Result<(String, usize), QueryError> {
    let mut index = start + 1;
    let mut value = Vec::new();
    let mut escaped = false;
    while index < input.len() {
        let byte = input[index];
        index += 1;
        if escaped {
            value.push(byte);
            escaped = false;
        } else if byte == b'\\' {
            escaped = true;
        } else if byte == b'"' {
            if value.is_empty() {
                return Err(QueryError::new(
                    "empty quoted phrase",
                    format!("empty quoted phrase at offset {start}"),
                ));
            }
            return Ok((String::from_utf8_lossy(&value).into_owned(), index));
        } else {
            value.push(byte);
        }
    }
    Err(QueryError::new(
        "unterminated quoted phrase",
        format!("unterminated quoted phrase at offset {start}"),
    ))
}

fn regex_literal(input: &[u8], start: usize) -> Result<(String, usize), QueryError> {
    let mut index = start + 1;
    let mut value = Vec::new();
    let mut escaped = false;
    while index < input.len() {
        let byte = input[index];
        index += 1;
        if escaped {
            value.push(b'\\');
            value.push(byte);
            escaped = false;
        } else if byte == b'\\' {
            escaped = true;
        } else if byte == b'/' {
            if value.is_empty() {
                return Err(QueryError::new(
                    "empty regular expression",
                    format!("empty regular expression at offset {start}"),
                ));
            }
            return Ok((String::from_utf8_lossy(&value).into_owned(), index));
        } else {
            value.push(byte);
        }
    }
    Err(QueryError::new(
        "unterminated regular expression",
        format!("unterminated regular expression at offset {start}"),
    ))
}

fn parse_filter(token: &str) -> Result<Option<(Field, String)>, QueryError> {
    let Some((field_name, value)) = token.split_once(':') else {
        return Ok(None);
    };
    if value.is_empty() {
        return Err(QueryError::new(
            "requires a value",
            format!("operator {:?} requires a value", format!("{field_name}:")),
        ));
    }
    let field = match field_name.to_ascii_lowercase().as_str() {
        "path" => Field::Path,
        "tag" => Field::Tag,
        "type" => Field::Type,
        "status" => Field::Status,
        "filename" => Field::Filename,
        "filetype" => Field::FileType,
        "created" => Field::Created,
        "modified" => Field::Modified,
        "index" | "index_status" => Field::Index,
        _ if operator_name(field_name) => {
            return Err(QueryError::new(
                "unknown search operator",
                format!("unknown search operator {:?}", format!("{field_name}:")),
            ));
        }
        _ => return Ok(None),
    };
    Ok(Some((field, value.to_owned())))
}

fn operator_name(value: &str) -> bool {
    let mut bytes = value.bytes();
    bytes.next().is_some_and(|byte| byte.is_ascii_alphabetic())
        && bytes.all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'_' | b'-'))
}

fn is_date_field(field: Field) -> bool {
    matches!(field, Field::Created | Field::Modified)
}

fn is_date_unit(value: &str) -> bool {
    matches!(
        value.to_ascii_lowercase().as_str(),
        "day" | "week" | "month" | "year"
    )
}

fn skip_space(input: &[u8], mut index: usize) -> usize {
    while index < input.len() && is_go_space(input[index]) {
        index += 1;
    }
    index
}

fn bare(input: &[u8], start: usize) -> (String, usize) {
    let mut index = start;
    while index < input.len() && !is_go_space(input[index]) {
        index += 1;
    }
    (
        String::from_utf8_lossy(&input[start..index]).into_owned(),
        index,
    )
}

fn is_go_space(byte: u8) -> bool {
    char::from(byte).is_whitespace()
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct DateRange {
    pub from: OffsetDateTime,
    pub to: OffsetDateTime,
}

/// Resolves one inclusive date expression against a fixed reference instant.
///
/// # Errors
///
/// Returns the Go-compatible validation error for invalid expressions.
pub fn parse_date_value(value: &str, reference: OffsetDateTime) -> Result<DateRange, QueryError> {
    let value = value.trim();
    if value.is_empty() {
        return Err(QueryError::new(
            "date filter requires a value",
            "date filter requires a value",
        ));
    }
    let reference = reference.to_offset(time::UtcOffset::UTC);
    let lower = value.to_ascii_lowercase();
    if lower.starts_with("last ") {
        let duration = match lower.as_str() {
            "last day" => Duration::days(1),
            "last week" => Duration::days(7),
            "last month" => Duration::days(30),
            "last year" => Duration::days(365),
            _ => {
                return Err(QueryError::new(
                    "invalid relative date",
                    format!("invalid relative date {value:?}"),
                ));
            }
        };
        return Ok(DateRange {
            from: reference - duration,
            to: reference,
        });
    }

    let parts: Vec<_> = value.split("..").collect();
    if parts.len() > 2
        || (parts.len() == 2 && (parts[0].trim().is_empty() || parts[1].trim().is_empty()))
    {
        return Err(QueryError::new(
            "invalid date range",
            format!("invalid date range {value:?}"),
        ));
    }
    let (mut from, from_date) = parse_date_endpoint(parts[0].trim())?;
    let (mut to, to_date) = if parts.len() == 2 {
        parse_date_endpoint(parts[1].trim())?
    } else {
        (from, from_date)
    };
    if from_date {
        from = from.date().with_time(Time::MIDNIGHT).assume_utc();
    }
    if to_date {
        to = to.date().with_time(Time::MIDNIGHT).assume_utc() + Duration::days(1)
            - Duration::nanoseconds(1);
    }
    if from > to {
        return Err(QueryError::new(
            "date range starts after it ends",
            format!("date range starts after it ends: {value:?}"),
        ));
    }
    Ok(DateRange { from, to })
}

/// Validates a comma-separated date value against the Unix epoch.
///
/// # Errors
///
/// Returns the first invalid-expression error.
pub fn validate_date_value(value: &str) -> Result<(), QueryError> {
    for expression in value.split(',') {
        if expression.trim().is_empty() {
            return Err(QueryError::new(
                "invalid date",
                format!("invalid date {value:?}"),
            ));
        }
        parse_date_value(expression, OffsetDateTime::UNIX_EPOCH)?;
    }
    Ok(())
}

fn parse_date_endpoint(value: &str) -> Result<(OffsetDateTime, bool), QueryError> {
    if let Ok(date) = Date::parse(value, DATE_FORMAT) {
        return Ok((date.with_time(Time::MIDNIGHT).assume_utc(), true));
    }
    if let Ok(timestamp) = OffsetDateTime::parse(value, &Rfc3339) {
        return Ok((timestamp.to_offset(time::UtcOffset::UTC), false));
    }
    Err(QueryError::new(
        "invalid date",
        format!("invalid date {value:?} (use YYYY-MM-DD or YYYY-MM-DD..YYYY-MM-DD)"),
    ))
}
